openshift-kube-apiserver/openshiftkubeapiserver/patch_handlerchain.go

```diff
@@ -69,15 +78,26 @@ func withOAuthInfo(handler http.Handler, oAuthMetadata []byte) http.Handler {
 
 // If we know the location of the asset server, redirect to it when / is requested
 // and the Accept header supports text/html
-func withConsoleRedirect(handler http.Handler, consolePublicURL string) http.Handler {
-	if len(consolePublicURL) == 0 {
+func withConsoleRedirect(handler http.Handler, c *genericapiserver.Config) http.Handler {
+	coreClient, err := corev1.NewForConfig(c.LoopbackClientConfig)
+	if err != nil {
+		klog.Errorf("unable to register /console redirect handler: %v", err)
 		return handler
 	}
 
 	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
 		if strings.HasPrefix(req.URL.Path, "/console") ||
 			(req.URL.Path == "/" && httprequest.PrefersHTML(req)) {
-			http.Redirect(w, req, consolePublicURL, http.StatusFound)
+			consolePublicConfig, err := coreClient.ConfigMaps(openShiftConfigManagedNamespaceName).Get(context.TODO(), consolePublicConfigMapName, v1.GetOptions{})
```

Comment:
this is not under auth or under p&f throttling, is it? So you can DoS the server (and etcd) by hammering on `/`.

Reply:
How often does this URL change? We could either watch the configmap, or we could have a very low-frequency poll loop.

Reply:
The URL changes rarely, but it is something an admin can configure.

Reply:
@sttts I am using ConfigMap informer so the etcd DoS shouldn't be a problem.

Regarding a p&f for `/`, there is no simple solution. There are these 3 options:
- move the hanlder behind genericapiserver.DefaultBuildHandlerChain: https://github.com/openshift/kubernetes/blob/d7b268fffba3e99a8fa283979bca093f55263e93/openshift-kube-apiserver/openshiftkubeapiserver/patch_handlerchain.go#L36
The default handler chain also includes auth so it would not be accessible. We could do the same as `.well-known` does and create a ClusterRole (or update system:openshift:public-info-viewer) that would grant the permission to outside traffic.

- we could inject the console handler when building the default handler chain here: https://github.com/openshift/kubernetes/blob/d7b268fffba3e99a8fa283979bca093f55263e93/staging/src/k8s.io/apiserver/pkg/server/config.go#L849 
- or we could inject the logic when installing the API: https://github.com/openshift/kubernetes/blob/d7b268fffba3e99a8fa283979bca093f55263e93/staging/src/k8s.io/apiserver/pkg/server/config.go#L930

We could also keep it as it is now, as there should not be a big overhead at the moment when processing a request, but it is not a great behaviour as we should rather respond with 429 status.

In any case I would like to remove the `/` root as it is too strong path for console IMO in an API server. And only works when preference for http is set. In my opinion `/console` should be enough.

Reply:
thoughts on these options and removal of `/` route?

Reply:
I don't think we need to proactively protect it with apf:
- the filter does thing only when the request URL prefix has `/console`
- there will be only a few requests starting with `/console`, (unless someone deliberately does a DoS attack with this, but it's no more vulnerable than the authn and authz filter that precede apf filter)
- this filter uses a lister to get the configmap